units/src/amount.rs

```diff
@@ -303,6 +355,86 @@ impl From<OutOfRangeError> for ParseAmountError {
     }
 }
 
+/// Error returned when the input string has higher precision than satoshis.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TooPreciseError {
+    position: usize,
+}
```

Comment:
+1 to adding a getter to anything that shows up in the display output. (If we don't, *somebody* is just going to parse it out of the string then complain if we ever change the string. Which we technically have the right to do, but it'd be better not to incentivize this sort of behavior.)

Reply:
Actually, and a bit sadly, I remember reading somewhere that the string output in `Display` is technically part of the API, I had forgotten until you mentioned it. That means by very strict standards we are never going to be able to change errors in a forward compatible way - perhaps we are waisting our time being so finicky?

Reply:
Perhaps we should have in the README of `1.0` crates
"We go to great pains to to comply with semver however due to the difficulty in doing so for errors we reserve the right to break your codes error paths if you are using the internals of our error types i.e., not using `anyhow`". 

And then just give up on the whole `non_exhaustive`, hider errors, ect. stuff?

One thing that is really bugging me is that we are painting ourselves into a place where we cannot modify any of our functions because of the returned error types - the only solution I've managed to come up with is having a "stable" marker like stdlib does for functions that we want to commit to, but I think this is going to far at the moment.

Reply:
In theory we can add more context and make the user call getters to access it, so we can improve our error ergonomics in that way. And if we really wanted to preserve the `Display` output we could add a wrapper type which would `Display` more data.

But honestly I think we'd be willing to bend the "cannot change `Display` without a major rev" rule if we found that there was additional data that we wanted to provide, or that our existing data was incorrect, or something.

Reply:
No, we definitely should not say "you may not use the internals or else we'll break you". If that's the case we should not expose the internals.

But I don't think that "hide the internals and provide accessors" is really painting ourselves into a corner. Especially if we're not exposing constructors so we're controlling all the contexts in which teh types are constructed.

Reply:
I'll make a discussions post so as not to hijack Kix's PR.

Reply:
WTF, I only saw this conversation now.

> is that definitely what we want?

Definitely not long term - I want to add spans. But I put it off for later since this is more important now.

> the string output in `Display` is technically part of the API, I had forgotten until you mentioned it. That means by very strict standards we are never going to be able to change errors in a forward compatible way - perhaps we are waisting our time being so finicky?

I don't believe it's reasonable to apply this to error messages. It certainly is for things like `Txid` though.

> And then just give up on the whole `non_exhaustive`, hider errors, etc?

Definitely not. Nobody sane  parses error messages but many people match on things or access them.



> One thing that is really bugging me is that we are painting ourselves into a place where we cannot modify any of our functions because of the returned error types

We can because we hide internals and use `non_exhaustive` - this very PR has the exact intention of allowing it!

> the only solution I've managed to come up with is having a "stable" marker like stdlib does for functions that we want to commit to, but I think this is going to far at the moment.

Yes, we should! I've figured it out a year ago when I realized we need to change the Taproot API based on real-world usage. And we should have similar condition for stabilizing: used in real-world applications without problems.